### DISCUSSION TITLE
WebGPU PipelineLayout question

### DISCUSSION THREAD

**Discussion by SetoKaiba** [emscripten-core/emscripten#17645] [Q&A] [answered]
How to set PipelineLayout to auto like the original js WebGPU does? Thanks.
@kainino0x

**Comment by kainino0x** [comment, marked as answer]
Setting the pipeline layout to `nullptr`/`NULL`/`0` should be equivalent to using `"auto"` in JavaScript.
https://github.com/emscripten-core/emscripten/blob/08bb133df6bcc3be226dd45d07887fa4d317408b/system/include/webgpu/webgpu.h#L1080

However Emscripten has not been updated to translate `0` into `"auto"` so right now you have to rely on the deprecated behavior in Chromium (assuming you're targeting Chromium). We plan to get Emscripten updated before removing the deprecated path from Chromium.

**Comment by SetoKaiba** [comment]


**Reply by kainino0x** [reply]
The automatic layout algorithm will only include entries in the bind group layout if they're "statically used" in the entry point (i.e. are mentioned in the shader code for the entry point or any function it calls). Is that the case for your shader?

**Reply by kainino0x** [reply]
Yes, the codebase is the same for all Chromium-based browsers as long as they have WebGPU support (Chrome, Chromium, Edge, etc.)

**Reply by kainino0x** [reply]
And yes, that shader statically uses the texture and sampler. I assume you are using pipeline.getBindGroupLayout to create the bind group?

It seems possible something is not getting passed through correctly to the JS API. Consider using the dev tools debugger to check the calls to createShaderModule, createRenderPipeline, createBindGroup to make sure they're getting the correct inputs.

**Comment by SetoKaiba** [comment]


**Reply by kainino0x** [reply]
What memory is growing? WASM heap memory, JS object memory, or something outside, perhaps in the browser engine or GPU process?

As long as you are dropping the wgpu:: objects, their destructors should be called and that should free memory in the WASM heap and in Emscripten on the JS side. However the Emscripten code is not that robust and might be leaking.

**Reply by kainino0x** [reply]
How are you measuring the memory usage increase? Try the dev tools memory profiling to investigate further if it's inside JS.

Since you had to write this code in JS (for now) you're doing a bit of "manual" memory management and it's possible that refs from C++ are not getting dropped properly.

**Reply by kainino0x** [reply]
If you can upload a pre-built version and serve it up on GitHub Pages maybe I can take a look. I don't have that much time to help though.

**Reply by kainino0x** [reply]
In Chrome Dev Tools' Memory panel, I'm seeing a consistent 19.0-19.3 MB memory usage, and the "allocation timeline" tracing is showing almost no allocations during runtime.

**Reply by kainino0x** [reply]
Oh, I forgot to do that. I tried again with that. I'm on Chrome Canary on Mac. No memory allocations seen by dev tools, which means I'm not seeing allocations from ArrayBuffer. Instead I'm seeing the memory usage of Chrome's GPU process going up, not at a consistent rate but does seem to grow indefinitely at times (have seen >100GB). It seems to grow mainly when Chrome is in the background. It seems like something is garbage collecting memory though because it will periodically drop back down.

That said, this memory doesn't seem to be real. My laptop only has 32GB of memory, and Mac's Activity Monitor reports low "Real Mem" usage:
<img width="661" alt="image" src="https://user-images.githubusercontent.com/606355/184999191-880b0c8e-7769-41d2-aa33-e355568f0232.png">
It sounds like a chrome issue, although it may be working-as-intended.

Now that you have this easy repro case (thank you!) can you file a bug against Chromium on crbug.com ? send the link here so I can make sure it's correctly triaged. Be sure to include system info from `chrome://gpu`.

**Reply by kainino0x** [reply]
The release in Emscripten is a little hidden. It is generated here:
https://github.com/emscripten-core/emscripten/blob/11053684521f6830fa4aa36bc79e0266cce0228d/src/library_webgpu.js#L38
which generates a C binding which calls mgrXXX.release. The C++ bindings call that C binding from their destructors.

However that's not relevant to GPUExternalTexture because it doesn't have a "mgr". In your code, the GPUExternalTexture isn't kept as a reference at all, so it should get GCed at any time after the function returns. Internally, it's held onto by the GPUBindGroup, but once the texture expires (which happens automatically), it should just be a lightweight shell with no heavy memory behind it.

**Reply by kainino0x** [reply]
GPUBindGroup doesn't hold a reference to the GPUExternalTexture object, in Emscripten or in the JS WebGPU API. The reference is internal (deep in the GPU process in Chromium). So the GPUExternalTexture JS object becomes unreachable the moment your `my_bind_group` function returns and can get GCed anytime after that.

**Reply by kainino0x** [reply]
Please file a bug. It may be the case that your code is holding GPUExternalTexture objects alive, which would be a bug in your code, but nonetheless this should not be causing 100GB of memory leak in Chromium's GPU process.

**Reply by SetoKaiba** [reply]
> [[active_imports]], of type [list](https://infra.spec.whatwg.org/#list)<GPUExternalTexture>
> A list of all imported textures which have not yet been closed. The items in list are checked during each animation frame to determine whether to close them.

https://gpuweb.github.io/gpuweb/#gpu-external-texture
I know the problem now. I should use `emscripten_set_main_loop(frame, 0, true);` instead of `emscripten_set_main_loop(frame, 60, true);`
But anyway, should there be a way to release GPUExternalTexture instead of dependency on requestAnimationFrame?

**Reply by kainino0x** [reply]
For reference for future readers, the bug is filed here: https://crbug.com/1353621
